### PR TITLE
Add initial code in order to process denunciation

### PR DESCRIPTION
**Pull request by sydhds**
* [ ] document all added functions
* [ ] try in sandbox /simulation/labnet
* [ ] unit tests on the added/changed features
  * [ ] make tests compile
  * [ ] make tests pass 
* [ ] add logs allowing easy debugging in case the changes caused problems
* [ ] if the API has changed, update the API specification

### CODE REVIEW

File: massa-execution-worker/src/speculative_roll_state.rs [hunk 1]
```diff
@@ -148,6 +149,63 @@ impl SpeculativeRollState {
...
+        let owned_count = self.get_rolls(addr);
+        let roll_to_slash = min(roll_count, owned_count);
+
+        // remove the rolls
+        let current_rolls = self
+            .added_changes
+            .roll_changes
+            .get_mut(addr)
+            .ok_or_else(|| {
+                ExecutionError::RuntimeError(format!("Cannot get roll changes for addr: {}", addr))
```
Comment (damip) [inline massa-execution-worker/src/speculative_roll_state.rs:173]: if the address does not exist anymore, let's not end into a runtime error, but simply don't slash anything

File: massa-execution-worker/src/speculative_roll_state.rs [hunk 2]
```diff
@@ -148,6 +149,63 @@ impl SpeculativeRollState {
...
+    /// * `amount`: number of deferred credits to slash
+    pub fn try_slash_deferred_credits(
+        &mut self,
+        slot: &Slot,
+        addr: &Address,
+        amount: &Amount,
+    ) -> Result<Amount, ExecutionError> {
+        let credits = self.get_address_deferred_credits(addr, *slot);
+
+        // Return an error here - this should never happen
```
Comment (damip) [inline massa-execution-worker/src/speculative_roll_state.rs:194]: if there are not enough deferred credits in "last_key_value", other deferred credits (not just the ones targeted to the latest slot) should be slashed.

Ideally we would do the following:
* loop over deferred credits by increasing slot S
  * try slashing the amount from the deferred credits targeting S
    * if there was enough to slash all the coins we needed to slash, break

**Review by damip — COMMENTED** [review]
Good ! A couple details (see comments) + missing documentation

File: massa-execution-worker/src/context.rs [hunk 3]
```diff
@@ -685,6 +685,72 @@ impl ExecutionContext {
...
+                )?;
+
+            slashed_coins = slashed_coins.saturating_add(slashed_coins_in_deferred_credits);
+            let amount_remaining_to_slash_2 =
+                slashed_coins.saturating_sub(slashed_coins_in_deferred_credits);
+            if amount_remaining_to_slash_2 > Amount::zero() {
+                warn!("Slashed {} coins (by selling rolls) and {} coins from deferred credits but cumulative amount is lower than expected: {} coins", 
+                    slashed_coins, slashed_coins_in_deferred_credits,
+                    self.config.roll_price.checked_mul_u64(roll_count)
+                        .ok_or_else(|| { ExecutionError::RuntimeError(format!("Cannot multiply roll price by {}", roll_count))})?
```
[the remark this answers is not in the document]
Comment (sydhds) [reply massa-execution-worker/src/context.rs]: Removed

File: massa-execution-worker/src/speculative_roll_state.rs (hunk 1 above)
Comment (sydhds) [reply massa-execution-worker/src/speculative_roll_state.rs]: Removed

File: massa-execution-worker/src/speculative_roll_state.rs (hunk 2 above)
Comment (sydhds) [reply massa-execution-worker/src/speculative_roll_state.rs]: Think I've fixed it :)

File: massa-execution-worker/src/speculative_roll_state.rs [hunk 4]
```diff
@@ -148,6 +149,61 @@ impl SpeculativeRollState {
...
+            let to_slash = min(*amount, remaining_to_slash);
+            let new_deferred_credits = amount.saturating_sub(to_slash);
+            remaining_to_slash = remaining_to_slash.saturating_sub(to_slash);
+
+            self.added_changes
+                .deferred_credits
+                .insert(*slot, *addr, new_deferred_credits);
+        }
+
+        remaining_to_slash
```
Comment (damip) [inline massa-execution-worker/src/speculative_roll_state.rs:204]: to make it uniform with try_slash_rolls, maybe let's make both functions return either the amount slashed or the remaining amount

File: massa-execution-worker/src/context.rs [hunk 5]
```diff
@@ -685,6 +685,69 @@ impl ExecutionContext {
...
+            let slashed_coins_in_deferred_credits = self
+                .speculative_roll_state
+                .try_slash_deferred_credits(&self.slot, denounced_addr, &amount_remaining_to_slash);
+
+            slashed_coins = slashed_coins.saturating_add(slashed_coins_in_deferred_credits);
+            let amount_remaining_to_slash_2 =
+                slashed_coins.saturating_sub(slashed_coins_in_deferred_credits);
+            if amount_remaining_to_slash_2 > Amount::zero() {
+                // Use saturating_mul_u64 to avoid an error (for a warn!)
+                warn!("Slashed {} coins (by selling rolls) and {} coins from deferred credits but cumulative amount is lower than expected: {} coins",
```
Comment (damip) [inline massa-execution-worker/src/context.rs:741]: add the address being slashed to the warning

**Comment by bilboquet** [comment]
Everything I could have think of has already been caught :)
looks good to me

File: massa-execution-worker/src/context.rs (hunk 5 above)
Comment (sydhds) [reply massa-execution-worker/src/context.rs]: Done

File: massa-execution-worker/src/context.rs [hunk 6]
```diff
@@ -709,6 +709,69 @@ impl ExecutionContext {
...
+                ))
+            })?
+            .saturating_sub(slashed_coins);
+
+        if amount_remaining_to_slash > Amount::zero() {
+            // There is still an amount to slash for this denunciation so we need to slash
+            // in deferred credits
+            let slashed_coins_in_deferred_credits = self
+                .speculative_roll_state
+                .try_slash_deferred_credits(&self.slot, denounced_addr, &amount_remaining_to_slash);
```
[the remark this answers is not in the document]
Comment (sydhds) [reply massa-execution-worker/src/context.rs:758]: Fixed by returning the amount slashed so this is consistent with the documentation

File: massa-execution-worker/src/speculative_roll_state.rs (hunk 4 above)
Comment (sydhds) [reply massa-execution-worker/src/speculative_roll_state.rs]: Fixed. See below

**Review by damip — APPROVED** [review]
Very good !

**Pull request merged**